### PR TITLE
Odyssey Stats: wire up purchase page

### PR DESCRIPTION
**Pull request by kangzj**
## Proposed Changes

The jump between WP Admin to Calypso blue is a well documented area of confusion for users. We get this comment all the time. So we are going to enable the page within Jetpack.

The PR only wire the page up, and doesn't handle anything else, so you'd expect some of the things not working, which would be fine at the stage.

- Adds an Odyssey implementation of `QueryProductsList`, which queries `jetpack/v4/products`
- Fixed a critical error for Odyssey implementation of `QuerySitePurchases`
- Wired up stats page in Odyssey Stats

Idea for future improvements

- We could add a mechanism to list the API and the origin of the requests, so that we don't have to override the components on build anymore

## Testing Instructions

* Build Jetpack if needed `jetpack build plugins/jetpack`
* Build Odyssey Stats `STATS_PACKAGE_PATH=~/path/to/jetpack/projects/packages/stats-admin yarn dev`
* Open `/wp-admin/admin.php?page=stats#!/stats/purchase/:siteSlug`
* Ensure you see the purchase page

image

**Note: the buttons to purchase Stats do not work at the moment and we'll fix those in follow up PRs.**

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

File: apps/odyssey-stats/src/components/odyssey-query-site-purchases.ts [hunk 1]
```diff
@@ -52,7 +52,7 @@ const useOdysseyQuerySitePurchases = ( siteId: number | null ) => {
 	}, [ siteId, reduxDispatch ] );
 };
 
-export default function OdysseyQuerySitePurchases( siteId: number | null ) {
+export default function OdysseyQuerySitePurchases( { siteId }: { siteId: number | null } ) {
```
Comment (kangzj) [inline apps/odyssey-stats/src/components/odyssey-query-site-purchases.ts:55]: ‼️

**Review by grzegorz-cp — APPROVED** [review]
👍

File: apps/odyssey-stats/src/components/odyssey-query-site-purchases.ts (hunk 1 above)
Comment (dognose24) [reply apps/odyssey-stats/src/components/odyssey-query-site-purchases.ts:55]: The `siteId` is used to [compare with existing](https://github.com/Automattic/wp-calypso/blob/trunk/client/state/purchases/reducer.js#L67C56-L67C69) fetched purchases, which worked in this case. 😅

File: apps/odyssey-stats/src/components/odyssey-query-products.ts [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+				// @ts-ignore
+				dispatch( receiveProductsList( productsList, type ) )
+			)
+			.catch( ( error: Error ) =>
+				dispatch( {
+					type: PRODUCTS_LIST_REQUEST_FAILURE,
+					error,
+				} )
+			);
+	}, [ dispatch, type, currency ] );
```
Comment (dognose24) [inline apps/odyssey-stats/src/components/odyssey-query-products.ts:46]:
```suggestion
		// eslint-disable-next-line react-hooks/exhaustive-deps
	}, [ dispatch, type, currency ] );
```

**Review by dognose24 — APPROVED** [review]
Works as described! 👍🏼

**Pull request merged**